File: android/src/main/java/com/nitroplay/streaming/HlsIdentity.kt
package com.nitroplay.hls

import com.margelo.nitro.video.NitroSourcePreviewConfig
import java.util.TreeMap

object HlsIdentity {
  fun requestKey(url: String, headers: Map<String, String>?): String {
    if (headers.isNullOrEmpty()) {
      return url
    }

    val stableHeaders = TreeMap(headers)
      .entries
      .joinToString("&") { (key, value) -> "$key=$value" }

    return "$url\n$stableHeaders"
  }

  fun previewKey(url: String, headers: Map<String, String>?, preview: NitroSourcePreviewConfig?): String {
    val profile = VideoPreviewProfile.from(preview)
    return buildString {
      append(requestKey(url, headers))
      append("\npreview:")
      append(profile.maxWidth)
      append('x')
      append(profile.maxHeight)
      append('@')
      append(profile.quality)
    }
  }
}

File: ios/streaming/HlsIdentity.swift
import Foundation

struct HlsIdentity {
  static func requestKey(url: String, headers: [String: String]?) -> String {
    guard let headers, !headers.isEmpty else {
      return url
    }

    let stableHeaders = headers
      .keys
      .sorted()
      .map { "\($0)=\(headers[$0] ?? "")" }
      .joined(separator: "&")

    return "\(url)\n\(stableHeaders)"
  }

  static func previewKey(url: String, headers: [String: String]?, profile: VideoPreviewProfile) -> String {
    "\(requestKey(url: url, headers: headers))\npreview:\(profile.maxWidth)x\(profile.maxHeight)@\(profile.quality)"
  }
}
